### PR TITLE
evidence: p2p refactor

### PR DESCRIPTION
**Pull request by alexanderbez**
## Description

Prepares the evidence reactor for the newly designed p2p changes per [ADR 062](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-062-p2p-architecture.md).

ref: #5670

### CODE REVIEW

File: evidence/reactor.go [hunk 1]
```diff
@@ -21,233 +46,322 @@ const (
...
-// SetEventBus implements events.Eventable.
-func (evR *Reactor) SetEventBus(b *types.EventBus) {
-	evR.eventBus = b
+// processPeerUpdate processes a PeerUpdate, returning an error upon failing to
+// handle the PeerUpdate or if a panic is recovered. For new or live peers it
+// will check if an evidence broadcasting goroutine needs to be started. For
+// down or removed peers, it will check if an evidence broadcasting goroutine
+// exists and signal that it should exit.
+//
+// FIXME: The peer may be behind in which case it would simply ignore the
```
Comment (alexanderbez) [inline evidence/reactor.go:210]: /cc @cmwaters :)

File: evidence/pool.go [hunk 2]
```diff
@@ -49,10 +49,9 @@ type Pool struct {
 // NewPool creates an evidence pool. If using an existing evidence store,
```
Comment (alexanderbez) [inline evidence/pool.go:49]: This file diff contains linting updates only.

File: p2p/peer.go [hunk 3]
```diff
@@ -130,6 +130,12 @@ func (puc *PeerUpdatesCh) Done() <-chan struct{} {
 	return puc.doneCh
 }
 
+// TestSend performs a blocking send of a PeerUpdate on the PeerUpdatesCh
+// updatesCh. Note, this should be used for testing purposes only.
+func (puc *PeerUpdatesCh) TestSend(pu PeerUpdate) {
```
Comment (alexanderbez) [inline p2p/peer.go:135]: @erikgrinaker Let me know how you feel about this.

It will probably be pretty common for isolated reactor tests to need the ability to directly send `PeerUpdate`s. I've added this function for testing purposes only (hence the name).

We can either:

1. Keep and use this method.
2. Turn `PeerUpdatesCh` into an interface (where we use a dummy/bidirectional channel for testing purposes).
3. Remove the read-only semantics from `PeerUpdatesCh`.

I prefer to stick with (1) or (2).

**Comment by alexanderbez** [comment]
I've attempted to leave the existing reactor's semantics unchanged as much as possible, however, since we no longer have access to peers directly, only their IDs, we cannot know of their height or if they're caught up or not.

Thus, the main change in semantics this PR introduces _currently_ is that evidence will _always_ be sent to a peer regardless if the recipient peer is able to accept/process that evidence or not. The old semantics in this context would simply just not gossip the evidence at all if it.

This results in the fact that we now send more messages over the wire. This yields the need for discussion around shared state, particularly around peers between reactors. However, this is a topic that should not block this PR and we can revisit it later.

File: p2p/peer.go (hunk 3 above)
Comment (erikgrinaker) [reply p2p/peer.go:135]: Why don't we just use the same patterns here as with `Channel`, i.e. take the channel in `NewPeerUpdatesCh`? Then the test can send on it all it wants.

Comment (alexanderbez) [reply p2p/peer.go:135]: Yeah I didn't think of that -- we can modify `NewPeerUpdates` to take a channel instead of internally constructing one.

File: evidence/reactor.go [hunk 4]
```diff
@@ -21,233 +46,346 @@ const (
...
+// handleEvidenceMessage handles enevelopes sent from peers on the EvidenceChannel.
+// It returns an error only if the Envelope.Message is unknown for this channel
+// or if the given evidence is invalid. This should never be called outside of
+// handleMessage.
+func (r *Reactor) handleEvidenceMessage(envelope p2p.Envelope) error {
+	switch msg := envelope.Message.(type) {
+	case *tmproto.EvidenceList:
+		r.Logger.Debug("received evidence list", "num_evidence", len(msg.Evidence), "peer", envelope.From.String())
+
+		// TODO: Refactor the Evidence type to not contain a list since we only ever
```
Comment (melekes) [inline evidence/reactor.go:132]: do we have an issue for this?

File: evidence/reactor.go [hunk 5]
```diff
@@ -21,233 +46,346 @@ const (
...
-		evR.Switch.StopPeerForError(src, err)
-		return
+// handleEvidenceMessage handles enevelopes sent from peers on the EvidenceChannel.
+// It returns an error only if the Envelope.Message is unknown for this channel
+// or if the given evidence is invalid. This should never be called outside of
+// handleMessage.
+func (r *Reactor) handleEvidenceMessage(envelope p2p.Envelope) error {
+	switch msg := envelope.Message.(type) {
+	case *tmproto.EvidenceList:
+		r.Logger.Debug("received evidence list", "num_evidence", len(msg.Evidence), "peer", envelope.From.String())
```
Comment (melekes) [inline evidence/reactor.go:130]: you can do `peerLogger := r.Logger.With("peer", envelope.From.String())` to create a temporary logger and use it throughout this func

**Review by melekes — APPROVED** [review]
👍

File: statesync/reactor.go [hunk 6]
```diff
@@ -350,6 +349,7 @@ func (r *Reactor) processSnapshotCh() {
 		select {
 		case envelope := <-r.snapshotCh.In():
 			if err := r.handleMessage(r.snapshotCh.ID(), envelope); err != nil {
+				r.Logger.Error("failed to process envelope", "ch_id", r.snapshotCh.ID(), "envelope", envelope, "err", err)
```
Comment (erikgrinaker) [inline statesync/reactor.go:352]: Shouldn't this be logged by the shim? We can just wrap the error with additional context.

File: evidence/pool.go (hunk 2 above)
Comment (erikgrinaker) [reply evidence/pool.go:49]: Thanks for the heads-up! 😅

File: evidence/reactor.go [hunk 7]
```diff
@@ -21,233 +46,346 @@ const (
...
+		evpool:       evpool,
+		evidenceCh:   evidenceCh,
+		peerUpdates:  peerUpdates,
+		closeCh:      make(chan struct{}),
+		peerRoutines: make(map[string]chan struct{}),
 	}
-	evR.BaseReactor = *p2p.NewBaseReactor("Evidence", evR)
-	return evR
+
+	evpool.SetLogger(logger)
```
Comment (erikgrinaker) [inline evidence/reactor.go:83]: Setup of injected dependencies should be done by the caller, e.g. it may already have set its own preferred logger for the evidence pool and we shouldn't override that with the reactor logger.

File: evidence/reactor.go [hunk 8]
```diff
@@ -21,233 +46,346 @@ const (
...
+	service.BaseService
+
+	evpool       *Pool
+	eventBus     *types.EventBus
+	evidenceCh   *p2p.Channel
+	peerUpdates  *p2p.PeerUpdatesCh
+	closeCh      chan struct{}
+	peerRoutines map[string]chan struct{}
+
+	mtx    tmsync.RWMutex
```
Comment (erikgrinaker) [inline evidence/reactor.go:62]: What does this mutex protect? It is idiomatic to place the mtx directly above the fields that are protected by it.

File: evidence/reactor.go [hunk 9]
```diff
@@ -21,233 +46,346 @@ const (
...
+// handleMessage.
+func (r *Reactor) handleEvidenceMessage(envelope p2p.Envelope) error {
+	switch msg := envelope.Message.(type) {
+	case *tmproto.EvidenceList:
+		r.Logger.Debug("received evidence list", "num_evidence", len(msg.Evidence), "peer", envelope.From.String())
+
+		// TODO: Refactor the Evidence type to not contain a list since we only ever
+		// send and receive one piece of evidence at a time. Or potentially consider
+		// batching evidence.
+		for i := 0; i < len(msg.Evidence); i++ {
```
Comment (erikgrinaker) [inline evidence/reactor.go:135]: nit: `for i, evidence := range msg.Evidence`, or alternatively `for i := range msg.Evidence`

File: evidence/reactor.go [hunk 10]
```diff
@@ -21,233 +46,346 @@ const (
...
+				// If we're given invalid evidence by the peer, notify the router that
+				// we should remove this peer by returning an error.
+				if _, ok := err.(*types.ErrInvalidEvidence); ok {
+					return err
+				}
+			}
+		}
+
+	default:
+		r.Logger.Error("received unknown message", "msg", msg, "peer", envelope.From.String())
```
Comment (erikgrinaker) [inline evidence/reactor.go:154]: This also gets logged in `processEvidenceChannel`, and should perhaps not even be logged there but rather in the shim?

File: evidence/reactor.go [hunk 11]
```diff
@@ -21,233 +46,346 @@ const (
...
+// processEvidenceCh implements a blocking event loop where we listen for p2p
+// Envelope messages from the evidenceCh.
+func (r *Reactor) processEvidenceCh() {
+	defer r.evidenceCh.Close()
+
+	for {
+		select {
+		case envelope := <-r.evidenceCh.In():
+			if err := r.handleMessage(r.evidenceCh.ID(), envelope); err != nil {
+				r.Logger.Error("failed to process envelope", "ch_id", r.evidenceCh.ID(), "envelope", envelope, "err", err)
```
Comment (erikgrinaker) [inline evidence/reactor.go:191]: I think we should use message as the external nomenclature here. Envelopes are just an internal addressing artifact, and isn't really a concept that users should know or care about.

File: evidence/reactor.go [hunk 12]
```diff
@@ -21,233 +46,346 @@ const (
...
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
+		// Check if we've already started a goroutine for this peer, if not we create
+		// a new done channel so we can explicitly close the goroutine if the peer
+		// is later removed, we increment the waitgroup so the reactor can stop
+		// safely, and finally start the goroutine to broadcast evidence to that peer.
+		_, ok := r.peerRoutines[peerUpdate.PeerID.String()]
+		if !ok {
+			r.peerRoutines[peerUpdate.PeerID.String()] = make(chan struct{})
+			r.peerWG.Add(1)
+			go r.broadcastEvidenceLoop(peerUpdate.PeerID, r.peerRoutines[peerUpdate.PeerID.String()])
```
Comment (erikgrinaker) [inline evidence/reactor.go:240]: Maybe use a local `doneCh` variable to make it clearer what the second argument is.

File: evidence/reactor.go [hunk 13]
```diff
@@ -21,233 +46,346 @@ const (
...
+// that the peer has already received or may not be ready for.
+//
+// REF: https://github.com/tendermint/tendermint/issues/4727
+func (r *Reactor) broadcastEvidenceLoop(peerID p2p.PeerID, doneCh <-chan struct{}) {
 	var next *clist.CElement
+
+	defer func() {
+		if e := recover(); e != nil {
+			r.Logger.Error("recovering from processing peer update panic", "err", e)
+			r.removePeerRoutine(peerID)
```
Comment (erikgrinaker) [inline evidence/reactor.go:299]: We may want to remove the peer routine here in all cases, since we want to guarantee that it's cleaned up when the function returns in all cases. Then we can just inline the logic as well, since we won't have to call it everywhere.

File: evidence/reactor.go [hunk 14]
```diff
@@ -21,233 +46,346 @@ const (
...
+			r.peerWG.Add(1)
+			go r.broadcastEvidenceLoop(peerUpdate.PeerID, r.peerRoutines[peerUpdate.PeerID.String()])
+		}
+
+	case p2p.PeerStatusDown, p2p.PeerStatusRemoved, p2p.PeerStatusBanned:
+		// Check if we've started an evidence broadcasting goroutine for this peer.
+		// If we have, we signal to terminate the goroutine via the channel's closure.
+		// This will internally decrement the peer waitgroup and remove the peer
+		// from the map of peer evidence broadcasting goroutines.
+		doneCh, ok := r.peerRoutines[peerUpdate.PeerID.String()]
```
Comment (erikgrinaker) [inline evidence/reactor.go:248]: This can end up closing an already closed channel, and thus panic. This happens e.g. if we receive another `PeerStatusDown` before the goroutine terminates (if the peer is strobing up/down or we receive duplicate updates).

One option would be to involve a `sync.Once` somehow. Another option would be to remove the entry here instead of during `removePeerRoutine`, in which case it's possible for multiple goroutines to briefly be running per peer (if we get another `Up` status before the previous goroutine terminates).

File: evidence/reactor.go [hunk 15]
```diff
@@ -21,233 +46,346 @@ const (
...
+		}
+	}()
+
+	r.Logger.Debug("received peer update", "peer", peerUpdate.PeerID.String(), "status", peerUpdate.Status)
+
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	switch peerUpdate.Status {
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
```
Comment (erikgrinaker) [inline evidence/reactor.go:231]: We probably want `New` to be a noop here -- it's a peer which we've discovered but not connected to yet.

**Review by erikgrinaker — COMMENTED** [review]
Looks great overall! Commented on some issues. I'll let the evidence guys vet the evidence handling logic, since they're more familiar with it.

File: evidence/reactor.go (hunk 7 above)
Comment (alexanderbez) [reply evidence/reactor.go:83]: I would rather change the pool's constructor API then -- this is a codesmell IMHO.

File: evidence/reactor.go (hunk 8 above)
Comment (alexanderbez) [reply evidence/reactor.go:62]: This is news to me. Can you show me some references that state that? It protects the `peerRoutines`.

File: evidence/reactor.go (hunk 9 above)
Comment (alexanderbez) [reply evidence/reactor.go:135]: `evidence` is bound to the scope of the loop. i.e. we need to index it. So I'd prefer to keep this as-is or use `for i := range msg.Evidence` but I don't see one as better than the other.

File: evidence/reactor.go (hunk 12 above)
Comment (alexanderbez) [reply evidence/reactor.go:240]: Sure thing!

File: evidence/reactor.go (hunk 8 above)
Comment (erikgrinaker) [reply evidence/reactor.go:62]: https://golang.org/doc/effective_go.html#commentary

> Grouping can also indicate relationships between items, such as the fact that a set of variables is protected by a mutex.

```go
var (
    countLock   sync.Mutex
    inputCount  uint32
    outputCount uint32
    errorCount  uint32
)
```

Comment (erikgrinaker) [reply evidence/reactor.go:62]: One convention that I recently discovered, and quite like, is to use an inline struct for mutex-protected fields -- this is used e.g. in CockroachDB:

```go
type Foo struct {
  logger log.Logger

  mu struct {
    sync.Mutex
    shared string
  }
}

func (f *Foo) Set(s string) {
  f.mu.Lock()
  defer f.mu.Unlock()
  f.mu.shared = s
}
```

File: evidence/reactor.go (hunk 9 above)
Comment (erikgrinaker) [reply evidence/reactor.go:135]: Oh, right -- taking pointers to loop variables is only dangerous if the references live on after the next loop iteration, which isn't the case here. But yeah, good catch anyway. Not a big deal.

File: evidence/reactor.go (hunk 8 above)
Comment (alexanderbez) [reply evidence/reactor.go:62]: Thanks!

File: statesync/reactor.go (hunk 6 above)
Comment (alexanderbez) [reply statesync/reactor.go:352]: Well I'd prefer to have it here for two reasons:

1. The shim is being removed, although the router will replace it, I don't know what it'll look like.
2. The shim isn't used in reactor tests and so when testing with verbosity enabled, these logs are very helpful (they helped me a lot when debugging). If you move them up to the shim, you lose them completely.

I suppose we could have more context-dependent logs in `handleMessage`. Would that work for you?

Comment (erikgrinaker) [reply statesync/reactor.go:352]: Good points. Yeah, I'm not really sure how we'll be doing this in the router, and whether it makes sense to log errors there (maybe only if they end up disconnecting the peer or something). Let's defer it for now.

File: evidence/reactor.go (hunk 10 above)
Comment (alexanderbez) [reply evidence/reactor.go:154]: There is no `processEvidenceChannel`. This is the only log that logs an unknown message.

Comment (erikgrinaker) [reply evidence/reactor.go:154]: Meant `processEvidenceCh`. We log this message `received unknown message` here, and also return an error `received unknown message` on the line below. That returned error is caught and logged here:

https://github.com/tendermint/tendermint/blob/d791336c0331e4477839f019d12a70d556b886f9/evidence%2Freactor.go#L190-L191

File: evidence/reactor.go (hunk 11 above)
Comment (alexanderbez) [reply evidence/reactor.go:191]: What about logging `"evidence", evidence`? I think we should log that as it contains useful information. Or we could log the To/From respectively.

Comment (erikgrinaker) [reply evidence/reactor.go:191]: Sure, makes sense

File: evidence/reactor.go (hunk 14 above)
Comment (alexanderbez) [reply evidence/reactor.go:248]: Yeah, sigh. Good catch. I wanted to explicitly remove it from the map in `broadcastEvidenceLoop` because that is what is responsible for message processing and handling the channel closure signal, otherwise, like you said we could have multiple running which is what I wanted to avoid. 

Let me think on this a bit...

File: evidence/reactor.go [hunk 16]
```diff
@@ -21,233 +46,346 @@ const (
...
+
+	for {
+		select {
+		case envelope := <-r.evidenceCh.In():
+			if err := r.handleMessage(r.evidenceCh.ID(), envelope); err != nil {
+				r.Logger.Error("failed to process envelope", "ch_id", r.evidenceCh.ID(), "envelope", envelope, "err", err)
+				r.evidenceCh.Error() <- p2p.PeerError{
+					PeerID:   envelope.From,
+					Err:      err,
+					Severity: p2p.PeerErrorSeverityLow,
```
Comment (cmwaters) [inline evidence/reactor.go:195]: What does a low severity peer error do/mean?

Comment (erikgrinaker) [reply evidence/reactor.go:195]: We're not really sure about the semantics here yet, we'll figure it out when we build the router. The idea is that a low severity might e.g. cause a disconnect, while a high severity might ban the peer entirely. But we might replace this with other concepts, or we might use different heuristics for this.

File: evidence/verify.go [hunk 17]
```diff
@@ -55,64 +76,100 @@ func (evpool *Pool) verify(evidence types.Evidence) error {
...
+			return types.NewErrInvalidEvidence(evidence, err)
+		}
+
+		return nil
 
 	case *types.LightClientAttackEvidence:
 		commonHeader, err := getSignedHeader(evpool.blockStore, evidence.Height())
 		if err != nil {
-			return err
+			return types.NewErrInvalidEvidence(evidence, err)
```
Comment (cmwaters) [inline evidence/verify.go:89]: This is just trying to retrieve a signed header from the store so we should just return a normal error instead of an invalid evidence one

File: evidence/verify.go [hunk 18]
```diff
@@ -55,64 +76,100 @@ func (evpool *Pool) verify(evidence types.Evidence) error {
...
 		}
+
 		trustedHeader := commonHeader
+
 		// in the case of lunatic the trusted header is different to the common header
 		if evidence.Height() != ev.ConflictingBlock.Height {
 			trustedHeader, err = getSignedHeader(evpool.blockStore, ev.ConflictingBlock.Height)
 			if err != nil {
-				return err
+				return types.NewErrInvalidEvidence(evidence, err)
```
Comment (cmwaters) [inline evidence/verify.go:103]: same here. This should return just a regular error

File: evidence/verify.go [hunk 19]
```diff
@@ -17,6 +17,11 @@ import (
 // - it is from a key who was a validator at the given height
 // - it is internally consistent with state
 // - it was properly signed by the alleged equivocator and meets the individual evidence verification requirements
+//
+// NOTE: Evidence may be provided that we do not have the block for validator
```
Comment (cmwaters) [inline evidence/verify.go:21]:
```suggestion
// NOTE: Evidence may be provided that we do not have the block or validator
```

File: evidence/verify.go [hunk 20]
```diff
@@ -25,26 +30,42 @@ func (evpool *Pool) verify(evidence types.Evidence) error {
...
-			state.LastBlockTime.Add(evidenceParams.MaxAgeDuration),
+		return types.NewErrInvalidEvidence(
+			evidence,
+			fmt.Errorf(
+				"evidence from height %d (created at: %v) is too old; min height is %d and evidence can not be older than %v",
+				evidence.Height(),
+				evTime,
+				height-evidenceParams.MaxAgeNumBlocks,
+				state.LastBlockTime.Add(evidenceParams.MaxAgeDuration),
+			),
```
Comment (cmwaters) [inline evidence/verify.go:68]: Just something to point out: You mentioned in a comment further up that we don't send invalid evidence when retrieving anything because we might be lagging behind the network and the evidence might be valid but we just don't have the blocks/validator set for it. In the opposite case where it is a peer that is lagging behind the node, the peer may send evidence that is still valid for them but no longer valid for the node in which case we will return an invalid evidence error. It's a pretty minor problem though.

File: evidence/reactor.go (hunk 4 above)
Comment (cmwaters) [reply evidence/reactor.go:132]: Once upon a time there was this: https://github.com/tendermint/tendermint/issues/4729

But I think following this PR, we might want to jot down all the potential improvements/changes that could be made

**Review by cmwaters — COMMENTED** [review]
looking good as far as I can tell

File: evidence/reactor.go (hunk 16 above)
Comment (cmwaters) [reply evidence/reactor.go:195]: Yeah I guess this is something else we can work out at a later point. It might not even be a static rule. Say for example if you're running low on active peers then you might still remain detected but in another circumstance where you are connected to too many then the slightest error could cause the router to disconnect.

In the case of this evidence reactor, we are already sifting through the more mild cases of errors such that this is almost only triggered when someone has been intentionally mischievous so most likely we should disconnect from them

Comment (erikgrinaker) [reply evidence/reactor.go:195]: Yeah, something like that -- the consensus reactor similarly "grades" peers as they do (or don't do) useful work. So we should probably have some sort of peer quality scheme that takes all of this into account.

File: evidence/reactor.go (hunk 4 above)
Comment (alexanderbez) [reply evidence/reactor.go:132]: Why was that closed @cmwaters? Would it still be beneficial to reference that issue in the code?

File: evidence/reactor.go (hunk 10 above)
Comment (alexanderbez) [reply evidence/reactor.go:152]: Ahh yes, I've tweaked the logging a bit.

File: evidence/reactor.go (hunk 4 above)
Comment (cmwaters) [reply evidence/reactor.go:130]: I believe we decided to go initially with individual evidence and then later we decided to generalise light client attack evidence and in doing so batch it together (although duplicate vote evidence is still individual).

Maybe we want to revisit this at a later point.

File: evidence/reactor.go (hunk 16 above)
Comment (alexanderbez) [reply evidence/reactor.go:192]: I'm actually going to leave this as `PeerErrorSeverityLow` for now until the router is implemented. Then we will revisit all the places we send `PeerError` to ensure the various severities make sense.

File: evidence/verify.go (hunk 17 above)
Comment (alexanderbez) [reply evidence/verify.go:89]: Nice catch -- updated.

File: evidence/verify.go (hunk 20 above)
Comment (alexanderbez) [reply evidence/verify.go:68]: A bit hard to follow that one. Are you saying we _should not_ return a `ErrInvalidEvidence` in this exact case? Currently, the peer would most likely disconnect in the current context. Should we just return a normal error here?

Comment (cmwaters) [reply evidence/verify.go:68]: I think this is a niche case and that we still should return `ErrInvalidEvidence` in order to avoid peers from spamming the node with expired evidence

File: evidence/pool.go [hunk 21]
```diff
@@ -48,11 +48,10 @@ type Pool struct {
 
 // NewPool creates an evidence pool. If using an existing evidence store,
 // it will add all pending evidence to the concurrent list.
-func NewPool(evidenceDB dbm.DB, stateDB sm.Store, blockStore BlockStore) (*Pool, error) {
-
+func NewPool(logger log.Logger, evidenceDB dbm.DB, stateDB sm.Store, blockStore BlockStore) (*Pool, error) {
```
Comment (cmwaters) [inline evidence/pool.go:51]: Are we actually adding this logger to the evidence pool though? Further down in the constructor we have
```go
...
logger:   log.NewNopLogger(),
...
```

Comment (cmwaters) [reply evidence/pool.go:51]: Also just out of curiosity, I would normally append a new argument to the end (or try order them from most important to least important). Is there a reason why you put it at the start?

File: evidence/reactor_test.go [hunk 22]
```diff
@@ -28,310 +25,535 @@ import (
...
+	primary.peerUpdatesCh <- p2p.PeerUpdate{
+		Status: p2p.PeerStatusNew,
+		PeerID: secondary.peerID,
 	}
 
-	p2p.MakeConnectedSwitches(config.P2P, N, func(i int, s *p2p.Switch) *p2p.Switch {
-		s.AddReactor("EVIDENCE", reactors[i])
-		return s
+	// The secondary reactor should have received all the evidence ignoring the
+	// already pending evidence.
```
Comment (cmwaters) [inline evidence/reactor_test.go:414]:
```suggestion
	// already committed evidence.
```

**Review by cmwaters — APPROVED** [review]
Just have one outstanding comment but other than that it looks good (at least from the evidence side) 🎉

File: evidence/reactor.go (hunk 4 above)
Comment (alexanderbez) [reply evidence/reactor.go:132]: Ok, I will still reference it then so future readers can have context.

File: evidence/reactor.go (hunk 14 above)
Comment (alexanderbez) [reply evidence/reactor.go]: I ended up opting for a simple modification to the current design -- `closer`, which simply wraps a `sync.Once` that ensures we close the channel only once. I've also added a test case for this.

I can imagine we might end up needing `closer` in other places, but for now I'll keep this private to the `evidence` pkg.

File: evidence/pool.go (hunk 21 above)
Comment (alexanderbez) [reply evidence/pool.go:51]: > Are we actually adding this logger to the evidence pool though? Further down in the constructor we have

🤦  good catch. I didn't see this from the original code. Thank you.

Comment (alexanderbez) [reply evidence/pool.go:51]: From my experience, objects like loggers, contexts and configurations come first.

But in general, arguments should be grouped by types and then by pointer semantics. But there isn't a hard rule to this AFAIK.

**Comment by alexanderbez** [comment]
bump @erikgrinaker

File: evidence/reactor.go (hunk 14 above)
Comment (erikgrinaker) [reply evidence/reactor.go]: Yeah, this pattern keeps popping up throughout the P2P stack, we may want a convenience wrapper for it somewhere. I wonder if we can maybe make use of a [`sync.Cond`](https://golang.org/pkg/sync/#Cond) for this kind of synchronization -- although I suppose they don't compose with channel selects.

File: evidence/reactor.go [hunk 23]
```diff
@@ -21,233 +46,363 @@ const (
...
-		}
-		evis[i] = ev
-	}
-
-	for i, ev := range evis {
-		if err := ev.ValidateBasic(); err != nil {
-			return nil, fmt.Errorf("invalid evidence (#%d): %v", i, err)
-		}
-	}
+func (r *Reactor) removePeerRoutine(peerID p2p.NodeID) {
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:402]: It was called multiple times before. I'll move it.

File: evidence/reactor.go [hunk 24]
```diff
@@ -21,233 +46,363 @@ const (
...
+
+	return nil
+}
+
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit.
+func (r *Reactor) OnStop() {
+	r.mtx.Lock()
+	for _, c := range r.peerRoutines {
+		c.close()
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:123]: Why would the router send peer updates (up/new) when reactors are shutting down?

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply evidence/reactor.go:123]: The router doesn't know when reactors shut down. If anything, the reactor could close the peer updates channel first, but then we'd have to make sure that peer updates channels are unbuffered and I think we'll want them to have a 1-size buffer at least (I'll explain why on the router PR, saw your comment on it).

[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:123]: In any case, this was done on purpose. Consider the case:

1. `broadcastEvidenceLoop` is blocking on `r.evidenceCh.Out() <- envelope`
2. Router stops (i.e `close(r.closeCh)`) which triggers the p2p channel to close via the `defer` in `processEvidenceCh`
3. A race condition occurs between trying to send (1) while it was closed (2).

[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:123]: > The router doesn't know when reactors shut down, and it has to keep running until all reactors are stopped.

How does it not know? It will have access to the routers, correct?

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply evidence/reactor.go:123]: > Router stops

Do you mean router or reactor? The router must run until all reactors have exited, we've set that as a basic requirement to avoid complicating reactor implementations.

File: evidence/reactor.go [hunk 25]
```diff
@@ -21,233 +46,363 @@ const (
...
 		select {
-		case <-afterCh:
-			// start from the beginning every tick.
-			// TODO: only do this if we're at the end of the list!
-			next = nil
-		case <-next.NextWaitChan():
-			// see the start of the for loop for nil check
-			next = next.Next()
-		case <-peer.Quit():
+		case r.evidenceCh.Out() <- envelope:
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:367]: I had this added on purpose. See comment https://github.com/tendermint/tendermint/pull/5747#discussion_r552086033. Essentially, we could block here while the router is stopping and the Channel is closed.

File: evidence/reactor.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply evidence/reactor.go:123]: > How does it not know? It will have access to the routers, correct?

No, it only has access to the channels and peer updates subscriptions, it doesn't know anything about reactors.

[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:123]: Reactor* stops.

File: evidence/reactor.go [hunk 26]
```diff
@@ -21,233 +46,363 @@ const (
...
+//
+// FIXME: The peer may be behind in which case it would simply ignore the
+// evidence and treat it as invalid. This would cause the peer to disconnect.
+// The peer may also receive the same piece of evidence multiple times if it
+// connects/disconnects frequently from the broadcasting peer(s).
+//
+// REF: https://github.com/tendermint/tendermint/issues/4727
+func (r *Reactor) processPeerUpdate(peerUpdate p2p.PeerUpdate) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:240]: That's true. Although, I think there is no harm in doing so. Why not protect against potentially faulty code in peer update handling?

But then again, now that I think about it, maybe we want to the panic to occur to point out the bug such that it can be fixed.

So yeah, I think it makes sense to remove this.

File: p2p/key.go [hunk 27]
```diff
@@ -36,6 +35,21 @@ func (id NodeID) Bytes() ([]byte, error) {
 	return bz, nil
 }
 
+// Equal returns true if two NodeIDs are equal.
+func (id NodeID) Equal(other NodeID) bool {
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply p2p/key.go:39]: Yes, where nodeIDs are used as accessors they should be normalized (we could have a constructor for this). 

But in this method, we're comparing the bytes, which are case insensitive (i.e. `A` == `a`)

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply p2p/key.go:39]: Sure, but if we normalize them then there's no point in converting to bytes, we can just compare the strings normally.

File: evidence/reactor.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply evidence/reactor.go:240]: Yeah, but that sort of goes for all code, why not put panic recovery everywhere? 🙂 I don't know, no harm in keeping it as you say, just trying to reduce boilerplate. Your call.

File: evidence/reactor.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply evidence/reactor.go:123]: > I think we need to ensure all broadcastEvidenceLoop have exited before calling close(r.closeCh) and so that's why I explicitly close all of them.

But why? The whole purpose of `closeCh` is to signal to goroutines to terminate, and we explicitly check that in `broadcastEvidenceLoop` -- why would we need to shut down goroutines before signalling them to shut down? 🙂

[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:123]: Because of the race condition between `defer r.evidenceCh.Close()` and `r.evidenceCh.Out() <- envelope`.

File: evidence/reactor.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:240]: I'd like to remove it because we could mask flaws. I'll handle the evidence reactor in this PR, but other reactors in another PR.

File: evidence/reactor.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply evidence/reactor.go:123]: Ah crap, you're right. Should've read your comments more carefully, sorry. We'll need to close the peer updates subscription first then, I think.

I don't think the select is necessary anyway though, since we'll have to wait for the waitgroup before closing.

File: p2p/key.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply p2p/key.go:39]: Yeah...how do you want to proceed then?

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply p2p/key.go:39]: Yeah, just use naive comparison for now -- that's what we're already doing with maps.

File: evidence/reactor.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:367]: Will apply this suggestion 👍

File: evidence/reactor.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply evidence/reactor.go:124]: > We'll need to close the peer updates subscription first then, I think.

We can't do this either because we use a single `closeCh` to signal to _all_ goroutines to exit (i.e. we don't have a unique `closeCh` per goroutine/domain). Although we could, I think it's not necessary and overkill.

Instead, I've opted for a simple `r.IsRunning()` check in `processPeerUpdate`.

File: p2p/key.go [hunk 28]
```diff
@@ -16,6 +16,11 @@ import (
 // NodeID is a hex-encoded crypto.Address.
 type NodeID string
 
+// NewNodeID returns a lowercased (normalized) NodeID.
+func NewNodeID(nodeID string) NodeID {
```
Comment (erikgrinaker) [inline p2p/key.go:20]: This should call `Validate()` as well, to make sure it's actually valid, and `Validate()` should check that it's lowercased (since callers may not be using `NewNodeID`, e.g. if it's received via Protobuf).

We should possibly have a `Normalize()` method as well, which can be called on already-typed strings received from external sourced.

**Review by erikgrinaker — APPROVED** [review]

File: evidence/reactor.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply evidence/reactor.go:124]: That'll work 👍

File: p2p/key.go (hunk 28 above)
Comment (alexanderbez) [reply p2p/key.go:20]: Why would this call `Validate` if it is doing the normalizing. If anything, we should just have `NewNodeID` and `Normalize`, where the former calls the later. Also, we could have a `Validate` function too.

Comment (erikgrinaker) [reply p2p/key.go:20]: Because it needs to e.g. check that it's actually a valid hex string, not just some random garbage.

`NodeID` can be embedded in structs, and these structs can be received via the network (e.g. from Protobuf handshakes), so we may want to have `Validate()` and/or `Normalize()` to handle this. I believe we also have some existing P2P code from the old stack that relies on `Validate()`.

Comment (alexanderbez) [reply p2p/key.go:20]: Ok, yeah I think having `NewNodeID` ensure input is a valid hex-string and then lower-case should suffice. We'll have a validate function as well.
